onmt/translate/translator.py

```diff
@@ -562,184 +548,81 @@ def _fast_translate_batch(
-                        alive_attn.size(0), -1, beam_size, alive_attn.size(-1))
-                    if alive_attn is not None else None)
-                non_finished_batch = []
-                for i in range(is_finished.size(0)):
-                    b = batch_offset[i]
-                    finished_hyp = is_finished[i].nonzero().view(-1)
-                    # Store finished hypotheses for this batch.
-                    for j in finished_hyp:
-                        hypotheses[b].append((
-                            topk_scores[i, j],
-                            predictions[i, j, 1:],  # Ignore start_token.
-                            attention[:, i, j, :memory_lengths[i]]
-                            if attention is not None else None))
-                    # End condition is the top beam finished and we can return
-                    # n_best hypotheses.
-                    if top_beam_finished[i] and len(hypotheses[b]) >= n_best:
-                        best_hyp = sorted(
-                            hypotheses[b], key=lambda x: x[0], reverse=True)
-                        for n, (score, pred, attn) in enumerate(best_hyp):
-                            if n >= n_best:
-                                break
-                            results["scores"][b].append(score)
-                            results["predictions"][b].append(pred)
-                            results["attention"][b].append(
-                                attn if attn is not None else [])
-                    else:
-                        non_finished_batch.append(i)
-                non_finished = torch.tensor(non_finished_batch)
-                # If all sentences are translated, no need to go further.
-                if len(non_finished) == 0:
-                    break
-                # Remove finished batches for the next step.
-                top_beam_finished = top_beam_finished.index_select(
-                    0, non_finished)
-                batch_offset = batch_offset.index_select(0, non_finished)
-                non_finished = non_finished.to(topk_ids.device)
-                topk_log_probs = topk_log_probs.index_select(0, non_finished)
-                batch_index = batch_index.index_select(0, non_finished)
-                select_indices = batch_index.view(-1)
-                alive_seq = predictions.index_select(0, non_finished) \
-                    .view(-1, alive_seq.size(-1))
-                if alive_attn is not None:
-                    alive_attn = attention.index_select(1, non_finished) \
-                        .view(alive_attn.size(0),
-                              -1, alive_attn.size(-1))
+            beam.advance(log_probs, attn)
+            if beam.is_finished.any():
+                beam.update_finished()
+            if beam.done:
+                break
+            select_indices = beam.current_origin
 
             # Reorder states.
             if isinstance(memory_bank, tuple):
```

Comment:
I wanted to make this change some time ago: source side vectors (here `memory_bank`, `memory_lengths`, and `src_map`) do not need to be reordered at each step, only when some batches are marked for removal.

Maybe this could be moved under `if beam.is_finished.any():`?

Reply:
Done. Oddly seems to be slower,
```
PRED AVG SCORE: -0.4390, PRED PPL: 1.5512
GOLD AVG SCORE: -11.2287, GOLD PPL: 75260.5316
Total translation time (s): 9.625983
Average translation time (s): 0.019252
Tokens per second: 1254.417403
```
but still a good change. (I think it's expected, but the decoder state needs mapped every step or the score changes)